File: 202-happy-number/happy-number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """

        num_visited = set()
      
        def next_number(n):
            num = 0
            while n != 0:
                rem = n % 10
                num = num + rem ** 2
                n = n / 10
            return num

        while n not in num_visited:
            num_visited.add(n)
            n = next_number(n)
            if n == 1:
                return True
            
        return False
        
        

File: 202-happy-number/happy-number.cpp
class Solution {

private:
    int next_num(int n) {
        int num = 0, rem = 0;

        while( n!= 0 ) {
            rem = n%10;
            num += rem*rem;
            n /= 10;
        }
        
        return num;
    }

public:
    bool isHappy(int n) {
        set<int> num_visited;

        while (num_visited.count(n) == 0) {
            num_visited.insert(n);
            n = next_num(n);

            if (n == 1) {
                return true;
            }
        }

        return false;
        
    }
};